### PR TITLE
4.0.29 candidate: EIP-27 implementation

### PR DESCRIPTION
**Pull request by kushti**
This PR contains implementation of EIP-27 soft-fork and its activation.

EIP-27 is described at https://github.com/ergoplatform/eips/pull/52/files , a copy is available in this PR in *emission.md* file. 

Please check the code carefully against the EIP

### CODE REVIEW

File: ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+  * Container for re-emission related contracts. Contains re-emission contract and pay-to-reemission contract.
+  */
+trait ReemissionContracts {
+
+  /**
+    * How much miner can take per block from re-emission contract
+    */
+  val reemissionRewardPerBlock: Long = 3 * 1000000000L // 3 ERG
```
Comment (pragmaxim) [inline ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala:19]: there is this `EmissionRules.CoinsInOneErgo` constant for this

File: src/main/scala/org/ergoplatform/reemission/ReemissionRules.scala [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    } else {
+      ErgoAddressEncoder.TestnetNetworkPrefix
+    }
+
+    implicit val addrEncoder = new ErgoAddressEncoder(networkPrefix)
+
+    val source =
+      """
+        |  {
+        |    INPUTS(0).value > 30000000L * 1000000000L
```
Comment (pragmaxim) [inline src/main/scala/org/ergoplatform/reemission/ReemissionRules.scala:66]: This seems to be a huge number, is that legit?

**Review by pragmaxim — APPROVED** [review]
Looks flawless, great job!

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala [hunk 3]
```diff
@@ -477,7 +497,7 @@ object ErgoWalletActor extends ScorexLogging {
     *
     * @param blockHeight - height of a block to scan
     */
-  private final case class ScanInThePast(blockHeight: ErgoHistory.Height)
+  private final case class ScanInThePast(blockHeight: ErgoHistory.Height, forceScan: Boolean)
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala:500]: missed ScalaDoc for forceScan

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala [hunk 4]
```diff
@@ -215,9 +218,17 @@ trait ErgoWalletSupport extends ScorexLogging {
...
-      val assets = changeBox.tokens.map(t => Digest32 @@ idToBytes(t._1) -> t._2).toIndexedSeq
-      new ErgoBoxCandidate(changeBox.value, changeAddressOpt.get, walletHeight, assets.toColl)
+    val changeBoxCandidates = selectionResult.changeBoxes.map { changeBoxAssets =>
+      changeBoxAssets match {
+        case candidate: ErgoBoxCandidate =>
+          candidate
+        case changeBox: ErgoBoxAssets =>
+          // todo: is this extra check needed ?
+          val reemissionTokenId = ergoSettings.chainSettings.reemission.reemissionTokenId
+          val assets = changeBox.tokens.filterKeys(_ != reemissionTokenId).map(t => Digest32 @@ idToBytes(t._1) -> t._2).toIndexedSeq
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala:228]: Please add comment on why reemissionTokenId token is filtered out?

File: src/main/scala/org/ergoplatform/reemission/ReemissionRules.scala (hunk 2 above)
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/reemission/ReemissionRules.scala:66]: Yes, this is 30M ERGs, see description in EIP-27

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala [hunk 5]
```diff
@@ -425,6 +431,10 @@ class ErgoTransactionSpec extends ErgoPropertyTest with ErgoTestConstants {
     }
   }
 
+  ignore("re-emission - valid flow") {
+    //todo: implement
+  }
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala:436]: Looks like missing important test.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

**Review by aslesarenko — CHANGES_REQUESTED** [review]

**Comment by aslesarenko** [comment]
@kushti what will happen when not enough votes and the activation height has been reached?

File: ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  lazy val payToReemission: ErgoTree = {
+    // output of the reemission contract
+    val reemissionOut = ByIndex(Outputs, IntConstant(0))
+
+    val rOutTokens = OptionGet(ExtractRegisterAs(reemissionOut, R2)(SCollection(STuple(SCollection(SByte), SLong))))
+
+    val firstTokenId = SelectField(ByIndex(rOutTokens, IntConstant(0)), 1.toByte)
+
+    EQ(firstTokenId, ByteArrayConstant(reemissionNftIdBytes))
+  }.toSigmaProp.treeWithSegregation
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala:44]: Also see #1700

File: src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala [hunk 7]
```diff
@@ -616,34 +617,99 @@ object CandidateGenerator extends ScorexLogging {
...
+          new Input(emissionBox.id, ProverResult.empty),
+          new Input(reemissionSettings.injectionBox.id, ProverResult.empty)
+        )
+      } else {
+        IndexedSeq(new Input(emissionBox.id, ProverResult.empty))
+      }
 
-      val minerBox = new ErgoBoxCandidate(emissionAmount, minerProp, nextHeight, assets)
+      val minerAmt = if (nextHeight == eip27ActivationHeight) {
+        // injection - inhection box value going to miner
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala:686]: ????

This is not minus but dash (minus does not make any sense here)

File: ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala (hunk 1 above)
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala:20]: Re-used it!

File: src/main/scala/org/ergoplatform/http/api/EmissionApiRoute.scala [hunk 8]
```diff
@@ -15,31 +17,61 @@ case class EmissionApiRoute(ergoSettings: ErgoSettings)
...
+    val minerReward = emissionRules.minersRewardAtHeight(height) - reemissionAmt
     val totalCoinsIssued = emissionRules.issuedCoinsAfterHeight(height)
     val totalRemainCoins = emissionRules.coinsTotal - totalCoinsIssued
-    EmissionInfo(minerReward, totalCoinsIssued, totalRemainCoins)
+    EmissionInfo(minerReward, totalCoinsIssued, totalRemainCoins, reemissionAmt)
   }
 
-  final case class EmissionInfo(minerReward: Long, totalCoinsIssued: Long, totalRemainCoins: Long)
-
+  // todo: add totalReemitted ?
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/http/api/EmissionApiRoute.scala:69]: What can be done later (without softfork ?) ? 
height added

File: src/main/scala/org/ergoplatform/settings/ReemissionSettings.scala [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  *
+  * @see src/main/resources/application.conf for parameters description
+  */
+case class ReemissionSettings(checkReemissionRules: Boolean,
+                              emissionNftId: ModifierId,
+                              reemissionTokenId: ModifierId,
+                              reemissionNftId: ModifierId,
+                              activationHeight: Int,
+                              reemissionStartHeight: Int,
+                              injectionBoxBytesEncoded: ModifierId) {
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/settings/ReemissionSettings.scala:20]: Changed to String

File: src/main/scala/org/ergoplatform/local/ErgoStatsCollector.scala [hunk 10]
```diff
@@ -158,7 +161,8 @@ object ErgoStatsCollector {
                       launchTime: Long,
                       lastIncomingMessageTime: Long,
                       genesisBlockIdOpt: Option[String],
-                      parameters: Parameters)
+                      parameters: Parameters,
+                      eip27Supported: Boolean)
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/local/ErgoStatsCollector.scala:187]: done for all the fields

File: ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/contracts/ReemissionContracts.scala]: done

**Comment by kushti** [comment]
> @kushti what will happen when not enough votes and the activation height has been reached?

Nothing will happened, new validation rules will not be activated.

File: src/test/scala/org/ergoplatform/settings/VotingSpecification.scala [hunk 11]
```diff
@@ -97,14 +97,28 @@ class VotingSpecification extends ErgoPropertyTest {
...
+    val esc = new ErgoStateContext(Seq(), None, ADDigest @@ Array.fill(33)(0: Byte), p, validationSettingsNoIl, vr, false)(updSettings)
+    val invalidVote = 100: Byte
+    val votes = Array(invalidVote , NoParameter, NoParameter)
+
+    // voting for non-existing param is okay if not start of an epoch
+    val h = defaultHeaderGen.sample.get.copy(height = 1, votes = votes, version = 0: Byte)
+    val esc2 = esc.appendHeader(h).get
+
+    val h2 = defaultHeaderGen.sample.get.copy(height = 2, votes = votes, version = 0: Byte)
+    esc2.appendHeader(h2).toEither.left.get.getMessage.contains("Incorrect vote") shouldBe true
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/test/scala/org/ergoplatform/settings/VotingSpecification.scala:112]: added

File: src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala [hunk 12]
```diff
@@ -57,7 +57,7 @@ class ErgoWalletSpec extends ErgoPropertyTest with WalletTestOps with Eventually
           log.info(s"Payment request $req")
           val tx = await(wallet.generateTransaction(req)).get
           log.info(s"Generated transaction $tx")
-          val context = new ErgoStateContext(Seq(genesisBlock.header), Some(genesisBlock.extension), startDigest, parameters, validationSettingsNoIl, VotingData.empty)
+          val context = new ErgoStateContext(Seq(genesisBlock.header), Some(genesisBlock.extension), startDigest, parameters, validationSettingsNoIl, VotingData.empty, false)
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala:60]: I do not see any value in that. Execution paths in the wallet do not depend on this flag.

**Comment by kushti** [comment]
@aslesarenko comments addressed, please make another pass!

File: src/main/resources/mainnet.conf [hunk 13]
```diff
@@ -25,6 +25,22 @@ ergo {
...
+
+      reemissionTokenId = "d9a2cc8a09abfaed87afacfbb7daee79a6b26f10c6613fc13d3f3953e5521d1a"
+
+      reemissionNftId = "d3feeffa87f2df63a7a15b4905e618ae3ce4c69a7975f171bd314d0b877927b8"
+
+      activationHeight = 777217
+
+      reemissionStartHeight = 2080800
+
+      injectionBoxBytesEncoded = "80a8d6b9071003040005808098f4e9b5ca6a0402d1ed91c1b2a4730000730193c5a7c5b2a47302008f9e2d0220fa2bf23962cdf51b07722d6237c0c7b8a44f78856c0f7ec308dc1ef1a92a5101d9a2cc8a09abfaed87afacfbb7daee79a6b26f10c6613fc13d3f3953e5521d1a808088fccdbcc32300fca71b8b95f6ad14ce600a126c8842334d40d35f8754176c4cda2c95219f19f700"
```
Comment (aslesarenko) [inline src/main/resources/mainnet.conf:42]: Does this bytes require update after ErgoTrees of the contracts became v1? (Note, the propertyBytes of ErgoTree does include the version)

Comment (kushti) [reply src/main/resources/mainnet.conf:42]: Injection box already in the chain, its script is v0. Will be used once only

File: src/main/resources/testnet.conf [hunk 14]
```diff
@@ -50,6 +50,22 @@ ergo {
...
+
+      reemissionTokenId = "01345f0ed87b74008d1c46aefd3e7ad6ee5909a2324f2899031cdfee3cc1e022"
+
+      reemissionNftId = "06f2c3adfe52304543f7b623cc3fccddc0174a7db52452fef8e589adacdfdfee"
+
+      activationHeight = 188001
+
+      reemissionStartHeight = 186400
+
+      injectionBoxBytesEncoded = "a0f9e1b5fb011003040005808098f4e9b5ca6a0402d1ed91c1b2a4730000730193c5a7c5b2a4730200f6ac0b0201345f0ed87b74008d1c46aefd3e7ad6ee5909a2324f2899031cdfee3cc1e02280808cfaf49aa53506f29034fb69b23d519f84c4811a19694b8cdc2ce076147aaa050276f0b840f40100325c3679e7e0e2f683e4a382aa74c2c1cb989bb6ad6a1d4b1c5a021d7b410d0f00"
```
Comment (aslesarenko) [inline src/main/resources/testnet.conf:66]: Same question about ErgoTree version change to v1.

Comment (kushti) [reply src/main/resources/testnet.conf:66]: This box is even spent in the testnet atm

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**